Implement Kibana/OpenSearch Dashboards dev services

Javatar81 opened the pull request:
<!--
If this is your first time contributing to the project, 
please consider reviewing https://github.com/quarkusio/quarkus/blob/main/CONTRIBUTING.md
-->

This change will add dashboard services to the dev services for extensions/elasticsearch-rest-client and extensions/elasticsearch-java-client. If the distribution is elasticsearch, it will start Kibana as the dashboard. If the distribution is opensearch, it will start OpenSearch-dashboards as the dashboard.

<!--
Please include in the description above the list of GitHub issues this Pull Request addresses in the following format:

See https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue
for more information about linking issues to the Pull Request.
-->

* Fixes #39582

Javatar81 committed:
First attmept to add Kibana to the dev services of the elastic extension

Javatar81 committed:
First working version

Javatar81 committed:
Enable / disable Kibana

Javatar81 committed:
Fully tested version of dev services for kibana and opensearch dashboards

Javatar81 committed:
Remove commented property

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:
```diff
@@ -244,6 +271,91 @@ private RunningDevService startElasticsearchDevServices(
+
+        if (!dockerStatusBuildItem.isContainerRuntimeAvailable()) {
+            log.warnf("Docker isn't working, please configure the Elasticsearch hosts property (%s).",
+                    displayProperties(buildItemConfig.hostsConfigProperties));
```
```suggestion
            log.warn("Docker is not working, cannot start the Kibana/OpenSearch dashboards dev service.");
```

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:
```diff
@@ -244,6 +271,91 @@ private RunningDevService startElasticsearchDevServices(
+                config.shared(),
+                launchMode.getLaunchMode())
+                .or(() -> ComposeLocator.locateContainer(composeProjectBuildItem,
+                        List.of(resolvedImageName.getUnversionedPart(), "elasticsearch", "opensearch"),
```
```suggestion
                        List.of(resolvedImageName.getUnversionedPart(), "kibana", "opensearch-dashboards"),
```

marko-bekhta reviewed (commented):
Hey 👋🏻 

thanks for the PR! 🙂

Could you please use rebase instead of adding merge commits? Something like:

```
git checkout main
git fetch upstream
git rebase upstream/main
```
where `upstream` points to this repo (`git remote add upstream https://github.com/quarkusio/quarkus.git` that is if you haven't set up the upstream yet) usually should do it for your current branch (`main` from which you've opened this PR)  🙂

I think it would be better if you just create another processor for the dashboards (i.e. `DevServicesElasticsearchDashboardsProcessor`), that way you'd easiely spot if you are not missing any stop/restart actions + the processor for the Elasticsearch services  is about to change to work with multiple services at the same time. 

You've made good progress here already 👍🏻 
Have you had a chance to look into this part about dev ui as well:

> it would be a great start if we provided either an option to launch Kibana in the Elasticsearch dev services (and include a link to it in the Dev UI),

it can go as a follow up if not 🙂

Javatar81 commented:
Thanks for all your comments. So I would move all my changes into a separate class DevServicesDashboardProcessor with its own properties files and address your proposed changes.

Regarding the Dev Ui. Yes I had that in mind but I planned it as a follow up.

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java

Fixed copy paste error when shutting down dashboard dev service as proposed by Marko

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java

Fixed copy paste error as proposed by Marko.

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Update build-parent/pom.xml

Fixed duplicate Kibana image prop as proposed by Marko.

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java

Fixed copy paste error as proposed by Marko.

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Renamed constant as proposed by Marko

Javatar81 committed:
Kibana / dashboards configured via hosts config if not empty.

gsmet commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/ElasticsearchCommonBuildTimeConfig.java:
```diff
@@ -133,9 +133,61 @@ interface ElasticsearchDevServicesBuildTimeConfig {
+             *
+             * @asciidoclet
+             */
+            @WithDefault(" ")
```
Small question: why do we pass an empty value here? Shouldn't it be an `Optional` instead?

gsmet reviewed (commented):
Thanks for this work, I see we are making good progress. I added a small question.

Javatar81 committed:
Make nodeOpts optional as proposed by Guillaume

Javatar81 committed:
Move dashboard devservices to separate processor class

Javatar81 committed:
Move dashboard devservice properties into separate properties files

Javatar81 commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:
```diff
@@ -296,6 +470,29 @@ private DockerImageName resolveImageName(ElasticsearchDevServicesBuildTimeConfig
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
```
Done

Javatar81 committed:
Use build items to propagate host and port of search backend deverservice to dashboards devservice

Javatar81 commented:
If anything else is needed, please let me know.

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+    private static final String DEV_SERVICE_KIBANA = "elasticsearch-kibana";
+    private static final String DEV_SERVICE_DASHBOARDS = "opensearch-dashboards";
+    private static final ContainerLocator elasticsearchContainerLocator = locateContainerWithLabels(ELASTICSEARCH_PORT,
+            DEV_SERVICE_LABEL, NEW_DEV_SERVICE_LABEL);
```
```suggestion
```

we can remove this one now, since we don't use it anymore for this processor

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+    private static final ContainerLocator dashboardContainerLocator = locateContainerWithLabels(DASHBOARD_PORT,
+            DEV_SERVICE_LABEL, NEW_DEV_SERVICE_LABEL);
+    private static final int LOCATE_BACKEND_MAX_RETRIES = 5;
+    private static final int LOCATE_BACKEND_INITIAL_WAIT_MILLIS = 5000;
```
```suggestion
```

Looks like these can go away, no ?

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+                    .map(host -> "http://" + host.replace("localhost", "host.docker.internal"))
+                    .collect(Collectors.toSet());
+        } else {
+            Optional<ContainerAddress> maybeContainerAddressSearchBackend = Optional.empty();
```
```suggestion
```
seems to be unused ...

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+        } else {
+            Optional<ContainerAddress> maybeContainerAddressSearchBackend = Optional.empty();
+            log.info(
+                    "no elasticsearch hosts config property found, using the host of theelasticsearch dev services container to connect");
```
```suggestion
            log.debug(
                    "No Elasticsearch hosts config property found, using the host of the Elasticsearch dev services container to connect");
```

marko-bekhta reviewed (commented):
hey 👋🏻 🙂 

sorry for not getting back to you sooner. Thanks for splitting out the Kibana processor! 
I've added a couple comments where we could cleanup a few things.

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

marko-bekhta commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+        }
+
+        DevservicesElasticsearchBuildItemsConfiguration buildItemsConfig = new DevservicesElasticsearchBuildItemsConfiguration(
+                devservicesElasticsearchBuildItems);
```
yes 👍🏻 🙂 , I was thinking since we've already processed that information once to start the requested distributions; and then based on those we've generated the `elasticsearchConnectionBuildItems`, we might as well add that extra bit of info and then we wouldn't need to do it one more time here, right?

Javatar81 replied:
If we potentially start two containers, it makes sense to split the DevServicesElasticsearchDashboardsProcessor into two separate processors, right?

Javatar81 committed:
Update extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java

Co-authored-by: Marko Bekhta <marko-bekhta@users.noreply.github.com>

Javatar81 committed:
Fixed safety belt and prepared dynamic lookup of hosts and distribution from elasticsearchConnectionBuildItems.

Javatar81 committed:
Support starting Dashboards and Kibana in parallel when both backends available either via hosts param or connection build items

Javatar81 committed:
Transform non-final static variables into instance variables

Javatar81 commented:
You can now start your application via `-Dquarkus.elasticsearch.hosts=localhost:9200,localhost:9100 -Dquarkus.elasticsearch.devservices.dashboard.enabled=true` and if localhost:9200 listens to a Elasticsearch instance and localhost:9100 to a Opensearch instance, the dev service will spin up a Dashboards and a Kibana container connected to the matching backend.

Javatar81 commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+            log.infof(
+                    "Dev Services for Elasticsearch Dashboards started. You can open th ui in your Browser at %s",
+                    getElasticsearchHosts(buildItemsConfig, devDashboardService));
+        }
```
Done in the new implementation.

Javatar81 commented on extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchDashboardsProcessor.java:
```diff
@@ -0,0 +1,303 @@
+        DockerImageName resolvedImageName = resolveDashboardImageName(config, resolvedDistribution);
+
+        Set<String> opensearchHosts;
+        if (buildItemConfig.hostsConfigProperties.stream().anyMatch(ConfigUtils::isPropertyNonEmpty)) {
```
> Thinking more about this ... hosts are a runtime property:

The hosts property is available via _DevservicesElasticsearchBuildItem_.

Javatar81 committed:
Small improvements in Javadocs for DashboardDevServicesConfig

Javatar81 committed:
Removed unused params as proposed by Marko

Javatar81 committed:
Renamed Strategy class to make clear that it is for the dashboard

Javatar81 committed:
Fix to resolve dashboard image instead of backend image.
